Cesium3DTilesContent/src/I3dmToGltfConverter.cpp

```diff
@@ -454,44 +455,75 @@ CesiumAsync::Future<ConvertedI3dm> convertI3dmContent(
         });
   }
   repositionInstances(decodedInstances);
-  AssetFetcherResult assetFetcherResult;
+  std::string baseUri;
   if (header.gltfFormat == 0) {
-    // Recursively fetch and read the glTF content.
-    auto gltfUri = std::string(
+    std::string gltfUri(
         reinterpret_cast<const char*>(gltfData.data()),
         gltfData.size());
-    return assetFetcher.get(gltfUri)
-        .thenImmediately(
-            [options, assetFetcher](AssetFetcherResult&& assetFetcherResult)
-                -> CesiumAsync::Future<GltfConverterResult> {
-              if (assetFetcherResult.errorList.hasErrors()) {
-                GltfConverterResult errorResult;
-                errorResult.errors.merge(assetFetcherResult.errorList);
-                return assetFetcher.asyncSystem.createResolvedFuture(
-                    std::move(errorResult));
-              }
-              return BinaryToGltfConverter::convert(
-                  assetFetcherResult.bytes,
-                  options,
-                  assetFetcher);
-            })
-        .thenImmediately([convertedI3dm = std::move(convertedI3dm)](
-                             GltfConverterResult&& converterResult) mutable {
-          if (converterResult.model)
-            convertedI3dm.gltfResult.model = std::move(converterResult.model);
-          convertedI3dm.gltfResult.errors.merge(converterResult.errors);
-          return convertedI3dm;
-        });
+    baseUri = CesiumUtility::Uri::resolve(assetFetcher.baseUrl, gltfUri);
   } else {
-    return BinaryToGltfConverter::convert(gltfData, options, assetFetcher)
-        .thenImmediately([convertedI3dm = std::move(convertedI3dm)](
-                             GltfConverterResult&& converterResult) mutable {
-          if (converterResult.model)
-            convertedI3dm.gltfResult.model = std::move(converterResult.model);
-          convertedI3dm.gltfResult.errors.merge(converterResult.errors);
-          return convertedI3dm;
-        });
-  }
+    baseUri = assetFetcher.baseUrl;
+  }
+  // This use of a lambda is the only way timoore knows to conditionally pass
+  // different futures to a then...() member function.
```

Comment:
I thought about using a ternary operator, but it's way less powerful than a lambda, being only an expression and all. Ultimately I didn't need to do anything too hairy inside this logic, but these days I kind of avoid the ternary operator in favor of lambda.

I had not thought about using std::move to call the next continuation "operator." I think we're getting off into matters of individual taste here; unless you feel strongly, I'm going to leave the current code as is.

Reply:
Well, I think continuation chains are a bit hard to read in the best of circumstances, and using a lambda here (we'd call this an IIFE - Immediately Invoked Function Expression - in JavaScript, but I don't know if C++ uses that term) makes it even more onerous than usual. clang-format really struggles with it. Another way that doesn't require a ternary or a separate function would be to use a std::optional to hold the Future, to work around the inability to default-construct one. Making Future default-constructible wouldn't be a terrible idea, either. A default-constructed Future would just be already rejected.

Anyway, I'm not _strongly_ opposed to it as-is.

Reply:
I ended up keeping the lambda but separating the definition from the invocation. If nothing else, it reduces the indentation of the continuations that follow by a lot, which is a pretty significant win IMO.

The optional thing would work but felt hacky. Making Future default constructible isn't as easy as I hoped because it would still need an AsyncSystem. And so there would be a bit of cost to default-constructing a thing that would never be used.